### PR TITLE
OneHot convertor update

### PR DESCRIPTION
**Pull request by IvanNovoselov**
Update arm_converter_one_hot due to revision of OneHot reference implementation (more details could be found in this [PR](https://github.com/openvinotoolkit/openvino/pull/4330))
Ticket: 37522

### CODE REVIEW

File: modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp [hunk 1]
```diff
@@ -2,88 +2,44 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include <details/ie_exception.hpp>
-
 #include "arm_converter/arm_converter.hpp"
 #include <ngraph/runtime/reference/one_hot.hpp>
 
 namespace ArmPlugin {
-template <typename Indices, typename OutputType>
-void wrap_one_hot(const Indices* arg,
```
Comment (l-bat) [inline modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp:11]: Why did you delete wrapper?

Comment (IvanNovoselov) [reply modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp:11]: I thought we don't need a function that does nothing but calls another function, so I fed ngraph::runtime::reference::xxx directly to MakeConversion(), like it is done [here](https://github.com/openvinotoolkit/openvino_contrib/blob/de5601166c0a5820fbf6f4644e7e3f74d921dcf8/modules/arm_plugin/src/arm_converter/arm_converter_cumsum.cpp#L18) or [here](https://github.com/openvinotoolkit/openvino_contrib/blob/de5601166c0a5820fbf6f4644e7e3f74d921dcf8/modules/arm_plugin/src/arm_converter/arm_converter_conv_backprop_data.cpp#L32) for example. But your question makes me think that I'm not getting it right. 
So could you please explain, why do we need the wrapper? To perform checks, like you mentioned in [this comment](https://github.com/openvinotoolkit/openvino_contrib/pull/52#discussion_r581815735)?

Comment (l-bat) [reply modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp:11]: We use wrappers to work with inconstant nodes. Instead of extraction a depth value at the graph construction stage, we pass an input node. The value will be extracted at the infer stage. We also need wrappers if `ngraph:: reference` has `char*` input value.

Comment (IvanNovoselov) [reply modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp:11]: Thank you! Will bring the wrapper back.

File: modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp [hunk 2]
```diff
@@ -2,88 +2,44 @@
...
-                                        one_hot_axis,
-                                        on_values[0],
-                                        off_values[0]);
-}
+    template<> Converter::Conversion::Ptr Converter::Convert(const opset::OneHot& node) {
+        const auto& out_shape = node.get_output_shape(0);
+        const std::int64_t axis = node.get_axis();
+        if (axis < 0 || axis >= out_shape.size())
+            THROW_IE_EXCEPTION << "Invalid axis value. Expected in [0, " << out_shape.size() <<
+                                "), got " << axis;
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp]: Negative axis handling is added.

File: modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp [hunk 3]
```diff
@@ -2,88 +2,44 @@
...
+        const auto& depth_const = std::dynamic_pointer_cast<opset::Constant>(node.input_value(1).get_node_shared_ptr());
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp:39]: Check is added.

**Comment by IvanNovoselov** [comment]
@l-bat please note that only int32 and int64 input types are supported by OneHot according to the updated [specs](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/sequence/OneHot_1.md). Is it Ok that the convertor supports [different types](https://github.com/IvanNovoselov/openvino_contrib/blob/9a176638bad3e025a668c0f672028c8e1a9fb3b9/modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp#L63-L100)?

**Comment by l-bat** [comment]
> @l-bat please note that only int32 and int64 input types are supported by OneHot according to the updated [specs](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/sequence/OneHot_1.md). Is it Ok that the convertor supports [different types](https://github.com/IvanNovoselov/openvino_contrib/blob/9a176638bad3e025a668c0f672028c8e1a9fb3b9/modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp#L63-L100)?

Let's add a check that the input types are `int32 `or `int64`, and remove the handling for the other types.

File: modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp [hunk 4]
```diff
@@ -2,88 +2,111 @@
...
-                  const OutputType* off_values) {
-    ngraph::runtime::reference::one_hot(arg,
-                                        out,
-                                        in_shape,
-                                        out_shape,
-                                        one_hot_axis,
-                                        on_values[0],
-                                        off_values[0]);
+                  const size_t out_elem_size,
+                  const std::int64_t depth,
```
Comment (l-bat) [inline modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp:13]:
```suggestion
                  const InputType* depth,
```

and extract the depth value below. With these changes, we will be able to work with inconstant depth.

File: modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp [hunk 5]
```diff
@@ -2,88 +2,111 @@
...
+                  const std::int64_t one_hot_axis,
+                  const OutputType* on_value,
+                  const OutputType* off_value) {
+    ngraph::runtime::reference::one_hot<InputType>(indices,
+                                                   indices_shape,
+                                                   reinterpret_cast<char*>(out),
+                                                   out_elem_size,
+                                                   static_cast<size_t>(depth),
```
Comment (l-bat) [inline modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp:21]:
```suggestion
                                                   static_cast<size_t>(depth[0]),
```

File: modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp [hunk 6]
```diff
@@ -2,88 +2,111 @@
...
+        if (depth != out_shape[axis])
+            THROW_IE_EXCEPTION << "Incompatible depth and axis values.";
+
+        auto make = [&] (auto refFunction) {
+            return MakeConversion(refFunction,
+                                  node.input(0),
+                                  ind_shape,
+                                  node.output(0),
+                                  node.get_output_element_type(0).size(),
+                                  depth,
```
Comment (l-bat) [inline modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp:56]:
```suggestion
                                  node.input(1),
```

Comment (IvanNovoselov) [reply modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp]: Thank you! Your suggestion was incorporated in the latest commit.

File: modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp (hunk 5 above)
Comment (IvanNovoselov) [reply modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp]: Thank you! Your suggestion was incorporated in the latest commit.

File: modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp (hunk 4 above)
Comment (IvanNovoselov) [reply modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp]: Thank you! Your suggestion was incorporated in the latest commit.

**Review by l-bat — APPROVED** [review]
Thank you 👍

**Comment by IvanNovoselov** [comment]
@l-bat thank you! Also, could you please delete the [OneHot_revision](https://github.com/openvinotoolkit/openvino_contrib/tree/OneHot_revision) branch? I pushed it by mistake, but have no permission to delete it.

**Comment by l-bat** [comment]
> @l-bat thank you! Also, could you please delete the [OneHot_revision](https://github.com/openvinotoolkit/openvino_contrib/tree/OneHot_revision) branch? I pushed it by mistake, but have no permission to delete it.

Done

**Comment by IvanNovoselov** [comment]
@l-bat some single-layer tests fail [Link to Jenkins](https://openvino-ci.intel.com/job/private-ci/job/ie/job/ie-tests-linux-debian_9_arm-cpu_arm/6545/testReport/). It appears to me that the problem is in the depth and indices type mismatch ([failing SLT source](https://github.com/openvinotoolkit/openvino/blob/799f34238ede58788d6f81f742d8260a992bb0c8/inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/one_hot.cpp#L45-L99)), so the types of depth and indices must be both integral but could still be different. Shall I revert the last commit and perform checks on depth in the converter?

**Comment by l-bat** [comment]
> @l-bat some single-layer tests fail [Link to Jenkins](https://openvino-ci.intel.com/job/private-ci/job/ie/job/ie-tests-linux-debian_9_arm-cpu_arm/6545/testReport/). It appears to me that the problem is in the depth and indices type mismatch ([failing SLT source](https://github.com/openvinotoolkit/openvino/blob/799f34238ede58788d6f81f742d8260a992bb0c8/inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/one_hot.cpp#L45-L99)), so the types of depth and indices must be both integral but could still be different. Shall I revert the last commit and perform checks on depth in the converter?

No, fix the wrapper (add new template argument)

**Comment by IvanNovoselov** [comment]
> > @l-bat some single-layer tests fail [Link to Jenkins](https://openvino-ci.intel.com/job/private-ci/job/ie/job/ie-tests-linux-debian_9_arm-cpu_arm/6545/testReport/). It appears to me that the problem is in the depth and indices type mismatch ([failing SLT source](https://github.com/openvinotoolkit/openvino/blob/799f34238ede58788d6f81f742d8260a992bb0c8/inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/one_hot.cpp#L45-L99)), so the types of depth and indices must be both integral but could still be different. Shall I revert the last commit and perform checks on depth in the converter?
> 
> No, fix the wrapper (add new template argument)

Implemented. Now the tests pass without fails ([latest tests](https://openvino-ci.intel.com/job/private-ci/job/ie/job/ie-tests-linux-debian_9_arm-cpu_arm/6617/testReport/)).

**Comment by azhogov** [comment]
CI passed: https://gitlab-icv.inn.intel.com/inference-engine/product-configs/merge_requests/1430
